core/src/main/java/org/apache/iceberg/OverwriteData.java

```diff
@@ -56,14 +67,34 @@ public OverwriteFiles addFile(DataFile file) {
   }
 
   @Override
-  public OverwriteFiles validateAddedFiles() {
-    this.validateAddedFiles = true;
+  public OverwriteFiles deleteFile(DataFile file) {
+    Preconditions.checkState(!overwriteByRowFilter,
+        "Either overwriteByRowFilter or deleteFile should be used, not both");
```

Comment:
I don't think these are mutually exclusive. I could still create an idempotent overwrite that deletes files directly.